File: core/src/com/alta_v2/physics/StateUpdater.java
package com.alta_v2.physics;

import com.alta_v2.physics.executionContext.TiledMapEngineContext;
import com.alta_v2.physics.executionContext.reserveData.ReservableActor;
import com.alta_v2.rendering.tiledMapScreen.TiledMapState;
import com.alta_v2.rendering.tiledMapScreen.state.NpcState;
import com.alta_v2.rendering.tiledMapScreen.state.PlayerState;
import lombok.experimental.UtilityClass;

import java.util.List;
import java.util.Map;

/**
 * Updates given state for new values from context.
 */
@UtilityClass
public class StateUpdater {

    public void updateAll(TiledMapState state, TiledMapEngineContext context) {
        updateTiledMapState(state, context);
        updatePlayerState(state.getPlayer(), context);
        for (NpcState npcState  : state.getNpcStateList()) {
            updateNpcState(npcState, context);
        }
    }

    private void updateTiledMapState(TiledMapState state, TiledMapEngineContext context) {
        state.updateMapCoordinates(context.getFocusPointGlobal().getX(), context.getFocusPointGlobal().getY());
    }

    private void updatePlayerState(PlayerState playerState, TiledMapEngineContext context) {
        playerState.updateCoordinates(
                context.getPlayer().getGlobalPoint().getX(), context.getPlayer().getGlobalPoint().getY()
        );
        playerState.setView(context.getPlayer().getView().getValue());
        playerState.setAnimationEnabled(context.getPlayer().getIsMoving().getValue());
        playerState.setAnimationChangeTime(context.getPlayer().getIsMoving().getChangeTime());
    }

    private void updateNpcState(NpcState npcState, TiledMapEngineContext context) {
        if (!context.getNpcMap().containsKey(npcState.getId())) {
            return;
        }

        npcState.getCoordinates().x = context.getNpcMap().get(npcState.getId()).getGlobalPoint().getX();
        npcState.getCoordinates().y = context.getNpcMap().get(npcState.getId()).getGlobalPoint().getY();
        npcState.setView(context.getNpcMap().get(npcState.getId()).getView().getValue());
    }

}

File: core/src/com/alta_v2/physics/StateUpdater.kt
package com.alta_v2.physics

import com.alta_v2.physics.executionContext.TiledMapEngineContext
import com.alta_v2.rendering.tiledMapScreen.TiledMapState
import com.alta_v2.rendering.tiledMapScreen.state.NpcState
import com.alta_v2.rendering.tiledMapScreen.state.PlayerState

/**
 * Updates given state for new values from context.
 */
class StateUpdater {

    companion object {
        fun updateAll(state: TiledMapState, context: TiledMapEngineContext) {
            updateTiledMapState(state, context)
            updatePlayerState(state.player, context)
            for (npcState in state.npcStateList) {
                updateNpcState(npcState, context)
            }
        }

        private fun updateTiledMapState(state: TiledMapState, context: TiledMapEngineContext) {
            state.updateMapCoordinates(context.focusPointGlobal.x, context.focusPointGlobal.y)
        }

        private fun updatePlayerState(playerState: PlayerState, context: TiledMapEngineContext) {
            playerState.updateCoordinates(
                    context.player.globalPoint.x, context.player.globalPoint.y
            )
            playerState.view = context.player.view.getValue()
            playerState.isAnimationEnabled = context.player.isMoving.getValue()
            playerState.animationChangeTime = context.player.isMoving.getChangeTime()
        }

        private fun updateNpcState(npcState: NpcState, context: TiledMapEngineContext) {
            if (!context.npcMap.containsKey(npcState.id)) {
                return
            }
            npcState.coordinates.x = context.npcMap[npcState.id]!!.globalPoint.x
            npcState.coordinates.y = context.npcMap[npcState.id]!!.globalPoint.y
            npcState.view = context.npcMap[npcState.id]!!.view.getValue()
        }
    }
}